proto/concept.proto

```diff
@@ -0,0 +1,963 @@
+        message Req {
+            bytes iid = 1;
+        }
+        message Res {
+            optional Attribute attribute = 1;
+        }
+    }
+
+    message GetSchemaExceptions {
+        message Req {}
+        message Res {
+            repeated Exception exceptions = 1;
+        }
+    }
+}
+
+message Exception {
+    string code = 1;
+    string message = 2;
+}
+
+message Concept {
+    oneof concept {
+        EntityType entity_type = 1;
+        RelationType relation_type = 2;
+        AttributeType attribute_type = 3;
+
+        RoleType role_type = 4;
+
+        Entity entity = 5;
+        Relation relation = 6;
+        Attribute attribute = 7;
+
+        ThingType.Root root_thing_type = 10;
+    }
+}
+
+message Thing {
+    oneof thing {
+        Entity entity = 1;
+        Relation relation = 2;
+        Attribute attribute = 3;
+    }
+
+    message Req {
+        bytes iid = 1;
+        oneof req {
+            Thing.Delete.Req thing_delete_req = 100;
+            Thing.GetHas.Req thing_get_has_req = 101;
+            Thing.SetHas.Req thing_set_has_req = 102;
+            Thing.UnsetHas.Req thing_unset_has_req = 103;
+            Thing.GetRelations.Req thing_get_relations_req = 104;
+            Thing.GetPlaying.Req thing_get_playing_req = 105;
+
+            Relation.AddPlayer.Req relation_add_player_req = 200;
+            Relation.RemovePlayer.Req relation_remove_player_req = 201;
+            Relation.GetPlayers.Req relation_get_players_req = 202;
+            Relation.GetPlayersByRoleType.Req relation_get_players_by_role_type_req = 203;
+            Relation.GetRelating.Req relation_get_relating_req = 204;
+
+            Attribute.GetOwners.Req attribute_get_owners_req = 300;
+        }
+    }
+
+    message Res {
+        oneof res {
+            Thing.Delete.Res thing_delete_res = 100;
+            Thing.SetHas.Res thing_set_has_res = 102;
+            Thing.UnsetHas.Res thing_unset_has_res = 103;
+
+            Relation.AddPlayer.Res relation_add_player_res = 200;
+            Relation.RemovePlayer.Res relation_remove_player_res = 201;
+        }
+    }
+
+    message ResPart {
+        oneof res {
+            Thing.GetHas.ResPart thing_get_has_res_part = 101;
+            Thing.GetRelations.ResPart thing_get_relations_res_part = 104;
+            Thing.GetPlaying.ResPart thing_get_playing_res_part = 105;
+
+            Relation.GetPlayers.ResPart relation_get_players_res_part = 202;
+            Relation.GetPlayersByRoleType.ResPart relation_get_players_by_role_type_res_part = 203;
+            Relation.GetRelating.ResPart relation_get_relating_res_part = 204;
+
+            Attribute.GetOwners.ResPart attribute_get_owners_res_part = 300;
+        }
+    }
+
+    message Delete {
+        message Req {}
+        message Res {}
+    }
+
+    message GetHas {
+        message Req {
+            oneof filter {
+                AttributeTypes attribute_types = 1;
+                Annotations annotation_filter = 2;
```

Comment:
Let's make an issue to remove the `oneof` here, it's not necessary IMO. We'll have to update server & clients with this change. Mark it as `good first issue` and assign it to Dmitrii M for future work.

Reply:
Note in the issue that the `repeated` can then be brought back inline.